### PR TITLE
Rename Adventure arguments

### PR DESCRIPTION
**Pull request by DerEchtePilz**
This essentially makes these changes to make arguments that deal with the same Minecraft argument type have identical names:

- Spigot-related argument changes:

```patch
- dev.jorel.commandapi.arguments.ChatArgument
- dev.jorel.commandapi.arguments.ChatColorArgument
- dev.jorel.commandapi.arguments.ChatComponentArgument

+ dev.jorel.commandapi.arguments.spigot.ChatArgument
+ dev.jorel.commandapi.arguments.spigot.ChatColorArgument
+ dev.jorel.commandapi.arguments.spigot.ChatComponentArgument
```

- Adventure-related argument changes:

```patch
- dev.jorel.commandapi.arguments.AdventureChatArgument
- dev.jorel.commandapi.arguments.AdventureChatColorArgument
- dev.jorel.commandapi.arguments.AdventureChatComponentArgument

+ dev.jorel.commandapi.arguments.adventure.ChatArgument
+ dev.jorel.commandapi.arguments.adventure.ChatColorArgument
+ dev.jorel.commandapi.arguments.adventure.ChatComponentArgument
```

TODO:
- [x] Replace instances of AdventureXArgument with XArgument
- [x] Update Kotlin DSL

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -369,14 +369,21 @@ This is the current roadmap for the CommandAPI (as of 11th May 2023):
         </tr>
     </thead>
     <tbody>
+        <tr>
+            <td valign="top"><b>10.0.0</b></td>
+            <td valign="top">?????????????????????????????</td>
+            <td valign="top">
+                <b>⚠️ This version is incompatible with any plugin that used component-related arguments in 9.x.x and below. (See documentation for more information)</b><br>
+            </td>
+        </tr>
```
Comment (DerEchtePilz) [inline README.md:378]: Probably fill this with more content.

File: README.md [hunk 2]
```diff
@@ -369,14 +369,21 @@ This is the current roadmap for the CommandAPI (as of 11th May 2023):
...
         <tr>
             <td valign="top"><b>9.0.4</b></td>
             <td valign="top">?????????????????????????????</td>
             <td valign="top">
                 <ul>
                     <li>Fixed the CommandAPI disabling datapacks on Paper 1.20.1 #40+ because it thought it was running on a Folia server</li>
                     <li>https://github.com/JorelAli/CommandAPI/pull/459 Added the ability to access raw arguments in the command executor</li>
-                    <li>https://github.com/JorelAli/CommandAPI/issues/469 Adds <code>AdventureChatColorArgument</code></li>
+                    <li>https://github.com/JorelAli/CommandAPI/issues/469 Adds <code>ChatColorArgument</code> for Adventure</li>
```
Comment (DerEchtePilz) [inline README.md:386]: Change this back to `AdventureChatColorArgument`

File: docssrc/src/upgrading.md [hunk 3]
```diff
@@ -1,5 +1,35 @@
 # Upgrading guide
 
+## From 9.?.? to 10.0.0
```
Comment (DerEchtePilz) [inline docssrc/src/upgrading.md:3]: Make this `9.?.?` say the correct version when this is ready to be merged

**Review by DerEchtePilz — COMMENTED** [review]
Yes, I did review my own PR :D
Basically, I just want to remind myself to actually make these changes when the time has come

**Comment by JorelAli** [comment]
The one mildly annoying thing with this PR is the developer experience. Developers now need to ensure that they import the "right one". There are no compile-time checks that will catch the case if a developer accidentally uses the wrong argument. (I'm only saying "mildly annoying", because the CommandAPI does have cases where it has classes which have the same name as other cases, such as NamespacedKey)

Not something worth rejecting the PR, just something to be aware of (of course, the fix for this is trivial on the developers point of view, and runtime errors will occur which should be detected during developer testing, so it's "easily identifiable"). We'll have to ensure that the documentation fully covers this to ensure that users don't mix them up (we should now have an intermediary "chat arguments" page, we can probably put a note about this there so developers know which package to import)

**Comment by willkroboth** [comment]
> The one mildly annoying thing with this PR is the developer experience. Developers now need to ensure that they import the "right one". There are no compile-time checks that will catch the case if a developer accidentally uses the wrong argument. (I'm only saying "mildly annoying", because the CommandAPI does have cases where it has classes which have the same name as other cases, such as NamespacedKey)

Yeah, it is a little confusing. Similar things happened to me when I accidentally import `java.awt.List` instead of `java.util.List`.

Maybe this could be similar to those `NamespacedKey` examples. There could be a `ChatArgument`/`ChatColorArgument`... class for the Spigot case and `ChatArgument.Adventure`/`ChatColorArgument.Adventure`... for Adventure. Or, since Spigot doesn't have to be the 'default', there could be `ChatArgument.Spigot`/`ChatColorArgument.Spigot`... instead.

Going that way would pretty much be no change at all though. If it's not a big change, it's probably not going to solve the problem you are trying to solve.

> Not something worth rejecting the PR, just something to be aware of (of course, the fix for this is trivial on the developers point of view, and runtime errors will occur which should be detected during developer testing, so it's "easily identifiable"). We'll have to ensure that the documentation fully covers this to ensure that users don't mix them up (we should now have an intermediary "chat arguments" page, we can probably put a note about this there so developers know which package to import)

As Skepter said, documentation could probably help with any possible confusion with the way it is right now.

**Comment by DerEchtePilz** [comment]
Alright, I guess, this should be done.

Once again, the only thing I am really not sure about is documentation. In this case, the `chatpreview.md` page as that refers to different platform implementation several times. I personally think I found a good compromise between only naming the class name and mentioning the platform every time too.

So yeah, a code review would be appreciated for the whole PR :D

File: docssrc/src/chatpreview.md [hunk 4]
```diff
@@ -22,13 +22,13 @@ For players that want to use chat preview, they must have `Chat Preview` enabled
...
 
-The `ChatArgument` and `AdventureChatArgument` classes include a method, `withPreview`:
+The `ChatArgument` classes include a method, `withPreview`:
 
 ```java
 public T withPreview(PreviewableFunction preview);
 ```
 
-The method `withPreview(PreviewableFunction preview)` lets you generate a preview to send to the client. This method takes in the `PreviewableFunction` functional interface, which is a function that takes in a `PreviewInfo` and returns either a `BaseComponent[]` (for `ChatArgument`) or a `Component` (for `AdventureChatArgument`):
+The method `withPreview(PreviewableFunction preview)` lets you generate a preview to send to the client. This method takes in the `PreviewableFunction` functional interface, which is a function that takes in a `PreviewInfo` and returns either a `BaseComponent[]` (for the Spigot-based `ChatArgument`) or a `Component` (for the Adventure-based`ChatArgument`):
```
Comment (willkroboth) [inline docssrc/src/chatpreview.md:31]: Just a tiny typo, it looks like there's a space missing. 
```suggestion
The method `withPreview(PreviewableFunction preview)` lets you generate a preview to send to the client. This method takes in the `PreviewableFunction` functional interface, which is a function that takes in a `PreviewInfo` and returns either a `BaseComponent[]` (for the Spigot-based `ChatArgument`) or a `Component` (for the Adventure-based `ChatArgument`):
```
Screen Shot 2023-09-25 at 10 10 27 AM

File: docssrc/src/chatpreview.md [hunk 5]
```diff
@@ -75,13 +75,13 @@ String fullInput();
 T parsedInput();
 ```
 
-`parsedInput()` is similar to `input()`, except it has been parsed by the CommandAPI's argument parser. This is a representation of what the argument in the executor would look like. For a `ChatArgument` the return type is `BaseComponent[]`, and for `AdventureChatArgument` the return type is `Component`.
+`parsedInput()` is similar to `input()`, except it has been parsed by the CommandAPI's argument parser. This is a representation of what the argument in the executor would look like. For the Spigot-based `ChatArgument` the return type is `BaseComponent[]`, and for the Adventure-based`ChatArgument` the return type is `Component`.
```
Comment (willkroboth) [inline docssrc/src/chatpreview.md:78]: Just a tiny typo, it looks like there's a space missing.
```suggestion
`parsedInput()` is similar to `input()`, except it has been parsed by the CommandAPI's argument parser. This is a representation of what the argument in the executor would look like. For the Spigot-based `ChatArgument` the return type is `BaseComponent[]`, and for the Adventure-based `ChatArgument` the return type is `Component`.
```
Screen Shot 2023-09-25 at 10 14 12 AM

**Review by willkroboth — CHANGES_REQUESTED** [review]
Just a few typos and small details I noticed. Otherwise, everything makes sense.

I'm still not sure what problem this is trying to solve. I've never used the chat arguments though, so I'm not sure how developers want to use them. One challenge I see with the setup here is providing support for Spigot and Adventure. Right now, if you want to use both `ChatArgument`s in the same file, you need to use the fully qualified name (eg. `dev.jorel.commandapi.arguments.spigot.ChatArgument`) for at least one of them, which doesn't seem convenient. I don't know if anyone would want to do that, but this is currently happening in the `ChatArgumentTests` files.

I think placing the arguments as static classes like `ChatArgument.Spigot` and `ChatArgument.Adventure` fits better with existing Arguments (Eg. `SoundArgument` and `SoundArgument.NamespacedKey` both use the `minecraft:resource_location` Brigadier argument). I don't know though. The best solution probably depends on the problem that this is trying to solve.

If we think different packages is the best solution,  this PR looks good to me :+1:.

**Comment by DerEchtePilz** [comment]
> Just a few typos and small details I noticed. Otherwise, everything makes sense.
> 
> I'm still not sure what problem this is trying to solve. I've never used the chat arguments though, so I'm not sure how developers want to use them. One challenge I see with the setup here is providing support for Spigot and Adventure. Right now, if you want to use both `ChatArgument`s in the same file, you need to use the fully qualified name (eg. `dev.jorel.commandapi.arguments.spigot.ChatArgument`) for at least one of them, which doesn't seem convenient. I don't know if anyone would want to do that, but this is currently happening in the `ChatArgumentTests` files.
> 
> I think placing the arguments as static classes like `ChatArgument.Spigot` and `ChatArgument.Adventure` fits better with existing Arguments (Eg. `SoundArgument` and `SoundArgument.NamespacedKey` both use the `minecraft:resource_location` Brigadier argument). I don't know though. The best solution probably depends on the problem that this is trying to solve.
> 
> If we think different packages is the best solution, this PR looks good to me 👍.

First, thanks for the review!
I am not really solving a problem with this since the arguments work fine the way they currently are, I just thought it would be nice if arguments that provide essentially the same functionality had the same name. [My initial discord message](https://discord.com/channels/745416925924032523/745419648970784821/1133075142340247632)
Of course, when I originally had the idea I didn't think about the case where someone would like to use both arguments in the same class (why would someone even want to do this, if you got Adventure, use Adventure).
Given that 10.0.0 isn't quite releasing any time soon, we still can discuss this package vs. inner class thing, I personally would continue to use this but using inner class would be fine as well.

**Comment by DerEchtePilz** [comment]
Superseded by #517

**Pull request closed without merging**